glossary.Rmd

```diff
@@ -0,0 +1,163 @@
+
+```{r}
+# Evaluated expression
+12 / 3
+
+# Quoted expression
+expr(12 / 3)
+```
+
+The technical definition of expressions is any R object that is created by [parsing](#glossary-parse) R code:
+
+* [Constants](#glossary-constant-symbolic) like `NULL`, `1`, `"foo"`, `TRUE`, `NA`, etc.
+* [Symbols](#glossary-sym) like `height` or `weight`
+* [Calls](#glossary-call) like `c()` or `list()`
+
+Unlike constants, symbols and calls are [symbolic](#glossary-constant-symbolic) objects: their [value](#glossary-eval) depends on the [environment](#glossary-env).
+
+
+### Expression (base)
+
+In base R, "expression" refers to a special type of vectors that contains quoted expressions in the rlang sense:
+
+```{r}
+base::expression(key <- "foo", toupper(key))
+```
+
+You'll most likely encounter this rare data structure as the return value of `base::parse()`:
+
+```{r}
+code <- "key <- 'foo'; toupper(key)"
+parse(text = code)
+```
+
+The only advantage of expression vectors compared to lists is that they include **source references**. The most apparent manifestation of source references is that expression vectors are printed with whitespace and comments preserved:
+
+```{r}
+code <- "{
+  # Interesting comment
+  weird     <-     whitespace
+}"
+parse(text = code, keep.source = TRUE)
+```
+
+Source references are mostly useful for debugging and development tools. They don't play any computational role and tidy evaluation doesn't make use of references. Consequently the parsing tools in rlang return normal lists of expressions (in the rlang sense) instead of expression vectors:
+
+```{r}
+rlang::parse_exprs(code)
+```
+
+
+### TODO Symbol {#glossary-sym}
+
+
+## Programming Concepts
+
+### Constant versus symbolic {#glossary-constant-symbolic}
+
+Constants, also called "literals", always have the same value no matter the context. On the other hand symbols and calls are [symbolic](#glossary-constant-symbolic) objects: their value depends on an [environment](#glossary-env) and what kind of objects are defined there.
+
+For instance the string `"mickey"` always represents the same string no matter the environment and what objects are defined there:
+
+```{r}
+# Here's a string:
+"mickey"
+
+mickey <- "mouse"
+
+# Still the same string:
+"mickey"
+```
+
+On the other [symbols](#glossary-sym) depend on current definitions:
+
+```{r}
+# We've defined `mickey` as "mouse"
+mickey
+
+mickey <- "mickey"
+
+# Now `mickey` is "mickey"
+mickey
+```
+
+One source of problem when you're working with quoted expressions is that they might be evaluated in arbitrary places, where objects have potentially been redefined to something completely different than expected. This is a common issue with tidyverse grammars because they evaluate quoted expressions in a [data mask](#glossary-data-mask). Say you'd like to divide a column by a factor defined in the current environment:
```

Comment:
One source of problems (plural). Maybe take out a couple of the modifiers here (e.g. just "something different than expected"